### ISSUE TITLE
P0849R8 `auto(x)`: `decay-copy` In The Language

### ISSUE THREAD

**Issue by StephanTLavavej** [microsoft/STL#2243]
[P0849R8](https://wg21.link/P0849R8) `auto(x)`: *`decay-copy`* In The Language

No feature-test macro is mentioned.

Despite being voted in as an LWG motion, this is a Core Language feature (which I've reported to the MSVC front-end devs). The changes to the Standard Library wording *appear* to be cleanups that don't affect observable behavior, but I haven't verified that. If this is indeed N/A for the Library, we should resolve this issue and remove the **cxx23** label, as it was never a (Library) feature to begin with.

**Comment by frederick-vs-ja** [comment]
I think P0849R8 may affect the observable behavior of `ranges::begin`, `ranges::end`, `ranges::rbegin`, and `ranges::rend`, by sometimes reduces one move-construction from a temporary iterator object.
`ranges::size` and `ranges::data` can be considered unchanged as move-construction from a temporary integer-like or pointer object is not observable.
Other functions are actually completely unchanged.

It seems that MSVC STL has implemented the new behaviors unconditionally (so does gcc libstdc++).

Perhaps we can consider the changes of CPOs to be DR even if WG21 doesn't say so, as the new behaviors are more reasonable and are what actually implemented.

**Comment by CaseyCarter** [comment]
> I think P0849R8 may affect the observable behavior of `ranges::begin`, `ranges::end`, `ranges::rbegin`, and `ranges::rend`, by sometimes reduces one move-construction from a temporary iterator object.

Yes, this is a case where the new tool gave us a better way to describe the intended behavior (which people were already implementing). There are, for example, cases where we said that `ranges::begin(meow)` was expression-equivalent to `decay-copy(meow.begin())`, which now require expression-equivalence to `auto(meow.begin())`. When `meow.begin()` is a glvalue, the two are equivalent expressions that yield a prvalue initialized from the glvalue. When `meow.begin()` is a prvalue, however, `decay-copy(meow.begin())` forces the prvalue to be materialized into an xvalue argument for `decay-copy` while `auto(meow.begin())` avoids materialization.

TLDR: the old wording really required implementations to do something like `auto operator()(range auto&& meow) { auto tmp = meow.begin(); return tmp; }` when we really intended something more like `auto operator()(range auto&& meow) { return meow.begin(); }`.

I suppose that means we should call this "already implemented" rather than "N/A", and annotate in `<yvals_core.h>` that we implement the proposal unconditionally in **C++20** mode.
